### ISSUE TITLE
Flow does not validate object property in if statement

### ISSUE THREAD

**Issue by jpolo** [facebook/flow#3974]
```javascript
function flags(): {| foo: boolean, bar: boolean |} {
	return {
      foo: true,
      bar: true,
    };
}

function foo() {
  const has = flags();
  const notExisting = has.notExisting; // is caught :-)
  
  if (has.notExisting) {/* is not caught :-( */ 
    alert('done')
  }
}
```

https://flow.org/try/#0GYVwdgxgLglg9mABMANgQwOYGcAUBKALkQG8AfZOOIgI0pQFM0wAaRatAJxrsaVIF8SAKACQHelBAckxIYnkLglIlA4h6zOQvnsuiVes3b+AbiH8hQ0JFgIKcfMPkQEWKIgAWaLIgC8ydGx8M2dXdzA4KABRAA8YNxgwDD9PbwA6COi4hKSTRAB6fMR4xAg0EAwPdwIAWjwtLRhgRBwvLAzI2PjYJLwSfIAqYp9M0vLK6pqcRAGirXk0Bg4oHAByABMEelX6+Qt+IA

**Comment by danwang** [comment]
Hey! It looks like JSX is a red herring, and that this is a feature, not a bug.

Here's the same behavior without JSX: ([flow try link](https://flow.org/try/#0MYewdgzgLgBAZgGwIYHMIwLwwBQEpMB8MA3gFAwwBOAplAK6VgnkXwggBcMUld1ANCwoAjJJS48+gigF8A3KXmlSoSLDjtMOfBiJkKq6DAAWSdFkSoIeBQfBHodOHC2mIAOjAgoAUQAeAJbQAWAoMAD8MADkwnTAANYw1GBIwgjUACZRMFwxcYkZQanpWQpKQA))

It looks like when a field is accessed in a ternary operator (or if predicate), it allows access but still doesn't know what type it is: ([flow try link](https://flow.org/try/#0MYewdgzgLgBAHgLhgbwIZLAVwLYCMCmATgL4wC8MAFFgDY1KpgCeAlANwBQHA9NzAKKFCIQgBoYqCDHxwADvmBR8AEw5wAdLi4BLAGZUNuFig4wYvGAGUo2ujGXgA5LADWYEAHcYHgBapYUEzyMLgw2hCm8JrqECDY+FA+2mAA5pzEQA))

**Comment by singpolyma-shopify** [comment]
does `const has : {| foo: bool, bar: bool |}` fix this?

**Comment by jpolo** [comment]
I think the intention for this feature was to enable access to a non existing property to test its value. IMHO, in a 'zero bug in production' type system, accessing something that does not exist is dangerous (even in an if condition) and should be caught by Flow. 

The `has.notExisting` expression should report an error unless refined by a `has.hasOwnProperty('notExisting')` or casted as a proper type `(has: Object).notExisting`.

**Comment by jpolo** [comment]
@singpolyma-shopify I tried in the sandbox, no it does change anything...

**Comment by Brianzchen** [comment]
Fixed in current latest version, tested 0.135.0
